lib/core-integration/src/Test/Integration/Scenario/API/ByronWallets.hs

```diff
@@ -188,7 +189,53 @@ spec = do
                     getFromResponse amount r0
             actualFee `shouldBe` predictedFee
 
-    describe "BYRON_ESTIMATE_06 - non-existing wallets" $  do
+    it "BYRON_MIGRATE_04 - \
+        \a migration operation removes all funds from the source wallet."
+        $ \ctx -> do
+
+            -- Restore a Byron wallet with funds, to act as a source wallet:
+            let sourceWalletName = "source wallet"
+            let sourceWalletPass = "source wallet passphrase"
+            sourceWallet <-
+                fixtureByronWalletWith sourceWalletName sourceWalletPass ctx
+
+            -- Verify that the source wallet has funds available:
+            r0 <- request @ApiByronWallet ctx
+                (getByronWalletEp sourceWallet) Default Empty
+            verify r0
+                [ expectResponseCode @IO HTTP.status200
+                , expectFieldSatisfy balanceAvailable (> 0)
+                ]
+
+            -- Perform a migration from the source wallet to a target wallet:
+            targetWallet <- emptyWallet ctx
+            r1 <- request @[ApiTransaction n] ctx
+                (migrateByronWalletEp sourceWallet targetWallet )
+                Default
+                (Json [aesonQQ|{"passphrase": #{sourceWalletPass}}|])
+            verify r1
+                [ expectResponseCode @IO HTTP.status202
+                , expectFieldSatisfy id (not . null)
```

Comment:
In the end, I think we can even be stronger in the assertion here. We expect exactly one transaction.

Reply:
> In the end, I think we can even be stronger in the assertion here. We expect exactly one transaction.

How can we be certain that there is exactly one transaction? Surely this depends on the UTxO distribution of the Byron fixture wallet? If the Byron fixture wallet has a complicated UTxO, that could easily result in more than one transaction being returned by `migrateByronWallet`.

If it is the case that the Byron fixture wallet has a very simple UTxO that causes only one transaction to be generated, surely that's an implementation detail that this test should not be concerned with?

Reply:
Fixture wallets always have the exact same distributions (fixed by the genesis configuration) :

10 UTxOs of 10M Ada (or 1M, not sure anymore).

Also, if you use fixtureWalletWith you can chose yourself what that distribution should be (the function takes a list of coins if I recall correctly).

Reply:
>Fixture wallets always have the exact same distributions

There is also this wallet in genesis -> https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/test/data/jormungandr/genesis.yaml#L68-L88. 

I have added this some time ago to have some variation... but I think that the utxo distribution is the same for it (i.e. 10 UTxOs), only the number of transaction is different on this particular one. (It has 1 tx, whereas others have 10txs). If it is problematic or confusing then maybe it should be changed.

Reply:
I think we should be cautious about making tests that depend on details of mock data, **unless** those details are exposed as shared, top-level, reusable constants.

While it may the case that migrating the default faucet wallet just creates a single transaction, I don't see any advantage of making our test depend on that fact.

That is, unless we define the default faucet UTxO as a reusable constant, and then make the number of expected transactions a function of that constant.

Reply:
> only the number of transaction is different on this particular one. 

But this particular wallet won't be exposed through a `faucet` like others, right? Unless the address generated for this wallet comes from one of the mnemonic used for the faucet (which would indeed be quite a wrong thing to do). If not, then there's nothing to worry about.

Reply:
> While it may the case that migrating the default faucet wallet just creates a single transaction, I don't see any advantage of making our test depend on that fact. I would argue that it's enough to ascertain that we have at least one transaction.

I'll introduce a `fixtureByronWalletWith` very much like its current dual for `fixtureWallet`. 
With this, we'll be able to make it obvious what the UTxO size is without having to rely on a implicit knowledge on the testing setup AND, more importantly, we'll able to test a scenario where the migration actually requires more than one transaction (by creating a big UTxO, very much we do for testing the limit of the fee policy).

Reply:
Yes, it is one of the faucet wallets. As far as I remember I just removed those first several `fund`s which just creates one inital tx with 10 utxos for this single wallet (as all those addresses belong to the same wallet)... others have 10 txs with 1 utxo each...

Reply:
Ah! It'd have been better not to use a faucet wallet for this, but to define the mnemonic separately :|

Reply:
> I think we should be cautious about making tests that depend on details of mock data

In this particular case one aspect is that we know what to expect with our fixture wallet mock (10 txs with 10 utxos)... unless someone silently breaks this expectation (like me with introducing this variation in genesis... which I think know is not really valuable, unless we really want to test this particular case... which we currently don't).

For that reason it is maybe safer (and clearer from the test reader perspective later on) to use `fixtureWalletWith` where we can define utxo distribution and even the number of transactions if needed.

Reply:
> For that reason it is maybe safer (and clearer from the test reader perspective later on) 

:100: